ListableUI/Sources/ListView/ListView.swift

```
@@ -740,6 +766,129 @@ public final class ListView : UIView
         }
     }
     
+    //
+    // MARK: Private - Scrolling
+    //
+    
+    private enum ScrollCompletionReason {
+        case cannotScroll
+        case scrolled(animated: Bool)
+    }
+    
+    /// This function is used by programmatic scrolling APIs that provide a scroll
+    /// completion handler. This will execute the `completion` handler after scrolling
+    /// is finished, or it will execute immediately if scrolling is not possible or if
+    /// animations are disabled.
+    private func handleScrollCompletion(reason: ScrollCompletionReason, completion: ScrollCompletion?) {
+        guard let completion else { return }
+        switch reason {
+        case .cannotScroll:
+            // Dispatch so that the completion handler executes on the next runloop
+            // execution.
+            DispatchQueue.main.async {
+                completion(ListStateObserver.DidEndScrollingAnimation(positionInfo: self.scrollPositionInfo))
+            }
+        case .scrolled(let animated):
+            if animated {
+                scrollCompletionHandler = completion
+            } else {
+                // Dispatch so that scrolling without an animation executes the closure
+                // on the next runloop execution, similar to scrolling with an animation.
+                DispatchQueue.main.async {
+                    /// Sync the `scrollPositionInfo` before executing the handler.
+                    self.performEmptyBatchUpdates()
+                    completion(ListStateObserver.DidEndScrollingAnimation(positionInfo: self.scrollPositionInfo))
+                }
+            }
+        }
+    }
+    
+    /// This is used to house the completion handler of scrolling APIs. This is kept
+    /// private and separate from `ListStateObserver` and its handlers.
+    private var scrollCompletionHandler: ScrollCompletion?
```

I think we want this to be an array we fill up then empty on run, right? Eg, what if someone enqueues two scroll events in quick succession, won't the first one get silently dropped? I think there's also some subtle behavior to consider here where really, only the last completion handler in that list will have fully completed its scroll event.

---

Thanks for the feedback! I haven't reproduced a case where the handlers can queue, but I'll do some more investigation - I'd like to capture that flow in a unit test if possible.

- [ ] Investigate handler queueing for a case that discards the original handler.

Here's what I've tested:
- If a user manually interrupts a `scrollTo(...)` animation by touching down in the scroll view, the completion handler will execute immediately, reporting the currently visible items.
- For programmatic interrupts, there are new unit tests in [`test_rapid_scroll_to_item_completion_calls()`](https://github.com/square/Listable/pull/582/files#diff-d4181dc057ced1cdc4d34137ca5e8eb4600b687ea0baad3ad37f80a339f9e974R1052) that capture the behavior of calls in quick succession:
   - Rapidly calling `scrollTo(...)` without an animation executes all handlers, reporting the final visible items.
   - When the target item is off screen, rapidly calling `scrollTo(...)` with an animation will execute all handlers, but only the handler from the most recent call to `scrollTo(...)` will report the item as visible.
   - When the target item is on screen, rapidly calling `scrollTo(...)` with an animation will execute all handlers and each handler will report that the item as visible.

---

I've pushed up an update to switch to a queue of scroll completion handlers. This also includes a test to simulate a handful of them being queued and executing at once.